File: SpaceInvaders/GameObject/Bomb/FallDagger.cs
using System;
using System.Diagnostics;

namespace SpaceInvaders
{
    class FallDagger : FallStrategy
    {
        public FallDagger()
        {
            this.oldPosY = 0.0f;
        }

        public override void Reset( float posY )
        {
            this.oldPosY = posY;
        }

        public override void Fall( Bomb pBomb )
        {
            Debug.Assert(pBomb != null);
            
            float targetY = oldPosY - 1.0f * pBomb.GetBoundingBoxHeight();

            if( pBomb.GetY() < targetY )
            {
                pBomb.MultiplyScale(1.0f, -1.0f);
                oldPosY = targetY;
            }
        }

        // Data
        private float oldPosY;
    }
}

File: quarkus-migration/space-invaders-quarkus/src/main/java/net/beeland/spaceinvaders/strategy/FallDagger.java
package net.beeland.spaceinvaders.strategy;

import net.beeland.spaceinvaders.gameobject.Bomb;

/**
 * FallDagger - Bomb falls in a dagger pattern by flipping vertically
 * Flips the bomb's Y scale when it reaches certain Y thresholds
 */
public class FallDagger implements FallStrategy {
    
    private float oldPosY;
    
    /**
     * Constructor
     */
    public FallDagger() {
        this.oldPosY = 0.0f;
    }
    
    @Override
    public void reset(float posY) {
        this.oldPosY = posY;
    }
    
    @Override
    public void fall(Bomb bomb) {
        if (bomb == null) {
            return;
        }
        
        // Calculate target Y position (one bounding box height below old position)
        float targetY = oldPosY - 1.0f * bomb.getBoundingBoxHeight();
        
        // If bomb has passed the target, flip vertically and update old position
        if (bomb.getY() < targetY) {
            bomb.multiplyScale(1.0f, -1.0f);
            oldPosY = targetY;
        }
    }
    
    /**
     * Get the old Y position
     * 
     * @return Old Y position
     */
    public float getOldPosY() {
        return this.oldPosY;
    }
}